chain/validation/state.go

```diff
@@ -20,80 +26,168 @@ import (
 
 func NewState() *StateWrapper {
 	bs := blockstore.NewBlockstore(datastore.NewMapDatastore())
 	cst := hamt.CSTFromBstore(bs)
+	// Put EmptyObjectCid value in the store. When an actor is initially created its Head is set to this value.
+	// Normally this happens in chain/vm/mkactor.go's init function. Without this flushing the state tree fails as this
+	// CID is not found in the store.
+	_, err := cst.Put(context.TODO(), map[string]string{})
+	if err != nil {
+		panic(err)
+	}
+
 	treeImpl, err := state.NewStateTree(cst)
 	if err != nil {
 		panic(err) // Never returns error, the error return should be removed.
 	}
+	root, err := treeImpl.Flush()
+	if err != nil {
+		panic(err)
+	}
 	storageImpl := &directStorage{cst}
-	return &StateWrapper{bs, newKeyStore(), treeImpl, storageImpl}
+	return &StateWrapper{bs, cst, newKeyStore(), root, storageImpl}
 }
 
 func (s *StateWrapper) Cid() cid.Cid {
-	panic("implement me")
+	return s.stateRoot
 }
 
 func (s *StateWrapper) Actor(addr vstate.Address) (vstate.Actor, error) {
 	vaddr, err := address.NewFromBytes([]byte(addr))
 	if err != nil {
 		return nil, err
 	}
-	fcActor, err := s.StateTree.GetActor(vaddr)
+	tree, err := state.LoadStateTree(s.cst, s.stateRoot)
+	if err != nil {
+		return nil, err
+	}
+	fcActor, err := tree.GetActor(vaddr)
 	if err != nil {
 		return nil, err
 	}
 	return &actorWrapper{*fcActor}, nil
 }
 
 func (s *StateWrapper) Storage(addr vstate.Address) (vstate.Storage, error) {
-	return s.directStorage, nil
+	return s.storage, nil
 }
 
 func (s *StateWrapper) NewAccountAddress() (vstate.Address, error) {
 	return s.keys.NewAddress()
 }
 
-func (s *StateWrapper) SetActor(addr vstate.Address, code cid.Cid, balance vstate.AttoFIL) (vstate.Actor, vstate.Storage, error) {
+func (s *StateWrapper) SetActor(addr vstate.Address, code vstate.ActorCodeCid, balance vstate.AttoFIL) (vstate.Actor, vstate.Storage, error) {
 	addrInt, err := address.NewFromBytes([]byte(addr))
 	if err != nil {
 		return nil, nil, err
 	}
+	lotusCode, err := fromActorCode(code)
+	if err != nil {
+		return nil, nil, err
+	}
+	tree, err := state.LoadStateTree(s.cst, s.stateRoot)
+	if err != nil {
+		return nil, nil, err
+	}
 	actr := &actorWrapper{types.Actor{
-		Code:    code,
+		Code:    lotusCode,
 		Balance: types.BigInt{balance},
+		Head:    vm.EmptyObjectCid,
 	}}
-
 	// The ID-based address is dropped here, but should be reported back to the caller.
-	// FIXME Lotus state tree requires the InitActor to be installed in order to resolve other
-	// actor addresses
-	_, err = s.StateTree.RegisterNewAddress(addrInt, &actr.Actor)
+	_, err = tree.RegisterNewAddress(addrInt, &actr.Actor)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, errors.Wrapf(err, "register new address for actor")
 	}
+	return actr, s.storage, s.flush(tree)
+}
 
-	_, err = s.StateTree.Flush()
+func (s *StateWrapper) SetSingletonActor(addr vstate.SingletonActorAddress, balance vstate.AttoFIL) (vstate.Actor, vstate.Storage, error){
```

Comment:
@anorth based on our out-of-band discussion about Singleton actor enums, if this matches your expectations I'll merge into `chainval` once this comment receives a :+1: